### ISSUE TITLE
[Feature Request] Support for winget

### ISSUE THREAD

**Issue by luxus** [rami3l/pacaptr#104]
hello,
i wonder if there is support for the official windows package manager
https://docs.microsoft.com/en-us/windows/package-manager/winget/
i prefer it over scoop and choco

**Comment by rami3l** [comment]
Thanks for making this feature request!

Actually, I have received the same request before, but at that time, it seemed that [`winget`](https://docs.microsoft.com/en-us/windows/package-manager/winget/) was not YET a full-featured package manager (it didn't even have `uninstall`)... That's why I decided to put it aside and wait.

When it reaches `v1.0` milestone (maybe later this year?), I'll be glad to return to this issue. Thanks :)

**Comment by luxus** [comment]
winget got released 16 days ago as 1.0 and included uninstall and upgrade.
the cmds are pretty easy for winget (and I use windows mostly for gaming) so its not so important for me have it supported with pacaptr 

https://github.com/microsoft/winget-cli/releases

**Comment by rami3l** [comment]
![telegram-cloud-photo-size-5-6235778398072974219-y](https://user-images.githubusercontent.com/33851577/122829322-24b43400-d2e7-11eb-91ed-430bb0d544d4.jpg)
Maybe I should wait for another while before trying to implement this 🤔

**Comment by luxus** [comment]
nobody expect it would be easy hehe.. but great that you tried... maybe we have to wait :D

**Comment by rami3l** [comment]
@ptanmay143 Thanks a lot for the gentle ping!

I have to confess that even though I made this project explicitly with Windows users in mind, I have moved away from Windows for quite a while personally (a bit before `winget` reached v1.0) , and I do fear of being irresponsible adding a half-baked solution to the release.

However, `winget` did get more interesting to look at in the previous year as it has been rapidly maturing. For example the previous issue [seems to be resolved](https://learn.microsoft.com/en-us/windows/package-manager/winget/uninstall#example-queries).

I am however willing to guide anyone interested through the process of adding such support to `pacaptr`, https://github.com/rami3l/pacaptr/pull/330 being a good example of it.

In short, the difficult part is to figure out how to perform the `pacman` operations could be mapped to subcommand invocations (to `winget` in this example) based on **real life usages**. I'll be glad if you're familiar enough with `winget` and are willing to figure that part out with me :)

**Comment by rami3l** [comment]
@ptanmay143 I've started playing around with my windows VM. So far the mapping should be quite straightforward:

- [x] `winget` is an `.exe`
- [ ] Q generates a list of installed packages
- [ ] Qc shows the changelog of a package
- [ ] Qe lists packages installed explicitly (not as dependencies)
- [ ] Qi displays local package information: name, version, description, etc
- [ ] Qii displays local packages which require X to be installed, aka local reverse dependencies
- [ ] Qk verifies one or more packages
- [ ] Ql displays files provided by local package
- [ ] Qm lists packages that are installed but are not available in any installation source (anymore)
- [ ] Qo queries the package which provides FILE
- [ ] Qp queries a package supplied through a file supplied on the command line rather than an entry in the package management database
- [ ] Qs searches locally installed package for names or descriptions
- [ ] Qu lists packages which have an update available
- [ ] R removes a single package, leaving all of its dependencies installed
- [ ] Rn removes a package and skips the generation of configuration backup files
- [ ] Rns removes a package and its dependencies which are not required by any other installed package, and skips the generation of configuration backup files
- [ ] Rs removes a package and its dependencies which are not required by any other installed package, and not explicitly installed by the user
- [ ] Rss removes a package and its dependencies which are not required by any other installed package
- [ ] S installs one or more packages by name
- [ ] Sc removes all the cached packages that are not currently installed, and the unused sync database
- [ ] Scc removes all files from the cache.
- [ ] Sccc (?)
- [ ] Sg lists all packages belonging to the GROUP
- [ ] Si displays remote package information: name, version, description, etc
- [ ] Sii displays packages which require X to be installed, aka reverse dependencies
- [ ] Sl displays a list of all packages in all installation sources that are handled by the packages management
- [ ] Ss searches for package(s) by searching the expression in name, description, short description
- [ ] Su updates outdated packages
- [ ] Suy refreshes the local package database, then updates outdated packages
- [ ] Sw retrieves all packages from the server, but does not install/upgrade anything
- [ ] Sy refreshes the local package database
- [ ] U upgrades or adds package(s) to the system and installs the required dependencies from sync repositories

**Comment by rami3l** [comment]
@luxus @ptanmay143 
It's very weird that `winget` is not a package manager currently avaiable in GitHub Actions which means I won't be able to add CI tests for it without manually locking to a version like in https://github.com/actions/runner-images/issues/910#issuecomment-1009640391.
